Add state option to git_config module for --unset

justinvhester opened the pull request:
##### SUMMARY
Fixes #28567

New option state: will accept value of set or unset. Calling the
git_config module with a config option defined by name: and state: unset
will remove that config option from git on the remote host. Tested on
Python 2.7.10 against a CentOS 7 VM. Tested to work with and without a
scope: defined. Also doesn't break (in my tests) if state: set. This patch
will simply apply the defined value: to the git config defined by name:.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
git_config

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (devel b69d17a862) last updated 2017/08/30 20:04:01 (GMT -500)
  config file = None
  configured module search path = [u'/Users/jvhester/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/jvhester/code/ansible-devel/lib/ansible
  executable location = /Users/jvhester/code/ansible-devel/bin/ansible
  python version = 2.7.10 (default, Oct 23 2015, 19:19:21) [GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
```

justinvhester committed:
Add state option to git_config module for --unset

New option state: will accept value of set or unset. Calling the
git_config module with a config option defined by name: and state: unset
will remove that config option from git on the remote host. Tested on
Python 2.7.10 against a CentOS 7 VM. Tested to work with and without a
scope: defined. Also doesn't break (in my tests) if state: set. This patch
will simply apply the defined value: to the git config defined by name:.

justinvhester committed:
Add version_added: "2.4"

justinvhester commented:
I think the issue type should have been "Feature Pull Request" instead of "Bugfix Pull Request".

willthames commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -157,9 +175,10 @@ def main():
             scope=dict(required=False, type='str', choices=['local', 'global', 'system']),
-            value=dict(required=False)
+            value=dict(required=False),
+            state=dict(required=False, choices=['set', 'unset'])
```
Use `present` or `absent` to match almost all other modules

justinvhester replied:
The choices have been updated to `present` and `absent`. I also altered the other bits of code affected by this change.

willthames commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -64,6 +64,12 @@
+  state:
+    description:
+      - Specify if a config should be C(set) or C(unset).
+    required: false
```
Add a `choices` key with `present` and `absent`.

justinvhester replied:
The `choices` key has been added.

justinvhester committed:
Update state options to 'present' and 'absent'

Taking the suggestion from reviewer, this commit changes the options for state to be consistent with almost all other Ansible modules.

justinvhester committed:
Update version_added to 2.5

djmattyg007 commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -64,6 +64,13 @@
         set that setting to the given value.
     required: false
     default: null
+  state:
```
All of the other parameters in this list are in alphabetical order. Please ensure the new `state` parameter is too.

justinvhester replied:
Understood, will rearrange in the next commit

djmattyg007 commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -157,9 +176,10 @@ def main():
             scope=dict(required=False, type='str', choices=['local', 'global', 'system']),
-            value=dict(required=False)
+            value=dict(required=False),
+            state=dict(required=False, choices=['present', 'absent'])
```
As above, all of the parameters in this list are in alphabetical order except `state`.

justinvhester replied:
Understood, will rearrange in the next commit. Thanks very much for the feedback.

djmattyg007 replied:
No problem. Congrats on your first Github PR :)

djmattyg007 commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -188,6 +208,11 @@ def main():
         new_value = None
 
+    if params['state'] == "absent":
+        config_state = params['state']
```
The name of this variable implies that it could also contain `"present"`, but it doesn't currently do that. I feel like it could be made clearer.

justinvhester replied:
I originally conceptualized something like this:
```python
if params['state'] == "absent":
    config_state = "--unset"
else:
    config_state = None
```
...since the default behavior of the `git_config` module is to set `name` to `value` without the need for `state: present`. If `config_state` turns out to be `present` then there is no change to the flags passed to `git config`. But if `config_state` is not None we know the user wants to `--unset`.

I'm all for more clarity though being new to the code I appreciate input on the best path forward.

This seems to offer less insight
```python
if params['state']:
    config_state = params['state']
```

This seems repetitive...
```python
if params['state'] == "absent":
    config_state = params['state']
elif params['state'] == "present":
    config_state = params['state']
else:
    config_state = None
```

djmattyg007 replied:
Maybe this?

    if params['state']:
        config_state = params['state']
    else:
        config_state = None

The AnsibleModule class has already verified that the value in params['state'] (if specified) is one of 'present' or 'absent', so you don't need to check that again yourself.

djmattyg007 commented on lib/ansible/modules/source_control/git_config.py:
```diff
@@ -209,10 +234,20 @@ def main():
+        # if the setting is present, insert --unset in args list and .run_command
+        old_val = out.rstrip()
+        args.insert(-1, "--unset")
+        (rc, out, err) = module.run_command(' '.join(args), cwd=dir)
```
This code does not account for check mode.

justinvhester replied:
Understood - I'll continue work on `module.check_mode` for this PR. Thanks again for the feedback.

djmattyg007 reviewed (changes requested):
Looks good so far, just needs a few small but important changes.

Akasurde commented:
@justinvhester Any news ?
